libraries/AP_HAL_Linux/Util_RPI.cpp

```diff
@@ -27,47 +27,67 @@ using namespace Linux;
 
 UtilRPI::UtilRPI()
 {
-    _check_rpi_version_by_rev();
+    _get_peripheral_base ();
 }
 
-int UtilRPI::_check_rpi_version_by_rev()
-{
-    // assume 2 if unknown
-    _rpi_version = 2;
-    const char *SOC[]= {"Broadcom BCM2835","Broadcom BCM2836","Broadcom BCM2837","Broadcom BCM2711"};
-    const char *revision_file_ = "/proc/device-tree/system/linux,revision";
-    uint8_t revision[4] = { 0 };
-    uint32_t cpu = 0;
-    FILE *fd;
 
-    if ((fd = fopen(revision_file_, "rb")) == NULL) {
-        printf("Can't open '%s'\n", revision_file_);
-    }
-    else {
-        if (fread(revision, 1, sizeof(revision), fd) == 4) {
-            cpu = (revision[2] >> 4) & 0xf;
-            
-            _rpi_version = cpu;
+// 
+// previous appraoch was using /proc/device-tree/system/linux,revision
+// now we use /proc/device-tree/soc/ranges see: https://forums.raspberrypi.com//viewtopic.php?t=244031
+uint32_t UtilRPI::_get_peripheral_base() 
+{
+    FILE *fp;
+    uint32_t base=0x00;
+    unsigned char buf[32];
+    _linux_board_version = LINUX_BOARD_TYPE::UNKNOWN_BOARD;
+    fp = fopen("/proc/device-tree/soc/ranges" , "rb");
+    if (fp) {
+        const uint16_t len = fread(buf, 1, sizeof(buf), fp);
+        if (len >= 8) {
+            base = buf[4]<<24 | buf[5]<<16 | buf[6]<<8 | buf[7];
+            if (!base)
+                base = buf[8]<<24 | buf[9]<<16 | buf[10]<<8 | buf[11];
+            if (!base)
+                base = buf[8]<<24 | buf[9]<<16 | buf[10]<<8 | buf[11];
+            if (!base && (len>15))
+                base = buf[12]<<24 | buf[13]<<16 | buf[14]<<8 | buf[15];
             
-            if (_rpi_version==0) {
-                _rpi_version=1; //RPI-Zero
-            }
-
-            printf("SOC: %s    use intern: %d \r\n",SOC[cpu], _rpi_version);
         }
-        else {
-            printf("Revision data too short\n");
-        }
-        fclose(fd);
+        fclose(fp);
     }
 
-    return _rpi_version;
-}
+    switch (base) {
+        case 0x0:
+            _linux_board_version = LINUX_BOARD_TYPE::UNKNOWN_BOARD;
+            printf("Cannot detect board-type \r\n");
+        break;
+        case 0x20000000:
+            _linux_board_version = LINUX_BOARD_TYPE::RPI_ZERO_1;
+            printf("RPI Zero / 1 \r\n");
+            printf("Prephiral base address is %x\n", base);
+        break;
+        case 0x3f000000:
+            _linux_board_version = LINUX_BOARD_TYPE::RPI_2_3_ZERO2;
+            printf("RPI 2, 3 or Zero-2 \r\n");
+            printf("Prephiral base address is %x\n", base);
```

Comment:
```suggestion
            printf("Perephiral base address is %x\n", base);
```

found a (4 actually) typo =]